pkg/reconciler/apis/apibinding/apibinding_reconcile_permissionclaim.go

```diff
@@ -0,0 +1,305 @@
+// It will also update the status if it finds an invalid permission claim.
+// Permission claims are considered invalid when the identity hashes are mismatched, and when there is no dynamic informer
+// for the group resource.
+func (c *controller) reconcilePermissionClaims(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
+	identityMismatch := false
+	lc := logicalcluster.From(apiBinding)
+
+	// Get all the bound resources in the locicalcluster
+	bindings, err := c.listAPIBindings(lc)
+	if err != nil {
+		return err
+	}
+	grsToBoundResource := map[apisv1alpha1.GroupResource]apisv1alpha1.BoundAPIResource{}
+	for _, binding := range bindings {
+		for _, resource := range binding.Status.BoundResources {
+			grsToBoundResource[apisv1alpha1.GroupResource{Group: resource.Group, Resource: resource.Resource}] = resource
+		}
+	}
+
+	invalidClaims := []apisv1alpha1.PermissionClaim{}
+	// get the added permission claims if added, add are just requeue the object.
+	addedPermissionClaims := mapGVRToClaim{}
+	for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+		found := false
+		for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+			if equality.Semantic.DeepEqual(acceptedPC, observedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			if acceptedPC.IdentityHash != "" {
+				if grsToBoundResource[acceptedPC.GroupResource].Schema.IdentityHash != acceptedPC.IdentityHash {
+					identityMismatch = true
+					invalidClaims = append(invalidClaims, acceptedPC)
+					continue
+				}
+			}
+			claim, gvr, err := c.createPermissionClaimHelper(acceptedPC)
+			if err != nil {
+				return err
+			}
+			claims := addedPermissionClaims[gvr]
+			if claims == nil {
+				claims = []permissionClaim{}
+			}
+			claims = append(claims, claim)
+			addedPermissionClaims[gvr] = claims
+		}
+	}
+
+	removedPermissionClaims := mapGVRToClaim{}
+	for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+		found := false
+		for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+			if equality.Semantic.DeepEqual(observedPC, acceptedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			claim, gvr, err := c.createPermissionClaimHelper(observedPC)
+			if err != nil {
+				return err
+			}
+			claims := removedPermissionClaims[gvr]
+			if claims == nil {
+				claims = []permissionClaim{}
+			}
+			claims = append(claims, claim)
+			removedPermissionClaims[gvr] = claims
+		}
+	}
+
+	klog.V(2).Infof("permission controller adding permission claims: %v, removing: %v and found invalid claims: %s", addedPermissionClaims, removedPermissionClaims, invalidClaims)
+
+	for gvr, claims := range addedPermissionClaims {
+		for _, claim := range claims {
+			// In the future, we will need to determine if the claim matches for any of the objects.
+			//TODO: use indexer instead of lister.
+			objs, err := claim.lister.Lister().List(labels.Everything())
```

Comment:
Please explain, is there a way to say that does not have this label, because then that would make sense, but added things won't have the label yet right?

Reply:
```go
requirement, err := labels.NewRequirement(claim.key, selection.DoesNotExist, []string{})
```

?